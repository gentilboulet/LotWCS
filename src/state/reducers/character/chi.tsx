import { Draft, produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../../actions/character/chi";
import { TChiState } from "../../models/character/chi";

import { increase } from "../../models/character/chi";

export type IChiAction = ActionType<typeof actions>;

export const chiReducer = produce(
  (draft: Draft<TChiState>, action: IChiAction) => {
    switch (action.type) {
      case getType(actions.chiBuy):
        increase(draft, action.payload.chi, action.payload.value);
        break;
      default:
    }
  },
);
