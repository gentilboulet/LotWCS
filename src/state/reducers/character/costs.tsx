import { Draft, produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../../actions/character/costs";
import { ICharacterState } from "../../models/character";
import { applyCost } from "../../models/character/costs";

export type ICostAction = ActionType<typeof actions>;

export const costReducer = produce(
  (draft: Draft<ICharacterState>, action: ICostAction) => {
    switch (action.type) {
      case getType(actions.payCost):
        applyCost(draft, action.payload.cost);
        break;
      default:
    }
  },
);
