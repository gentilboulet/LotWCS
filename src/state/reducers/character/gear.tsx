import { Draft, produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../../actions/character/gear";
import { TGearState } from "../../models/character/gear";

export const gearReducer = produce(
  (draft: Draft<TGearState>, action: ActionType<typeof actions>) => {
    switch (action.type) {
      case getType(actions.armorBuy):
        draft.armors.push({
          description: action.payload.description,
          name: action.payload.name,
          type: action.payload.type,
        });
        break;
      case getType(actions.armorCustomize):
        draft.armors[action.payload.armorIdx].description =
          action.payload.description;
        break;
      case getType(actions.weaponBuy):
        draft.weapons.push({
          description: action.payload.description,
          name: action.payload.name,
          type: action.payload.type,
        });
        break;
      case getType(actions.weaponCustomize):
        draft.weapons[action.payload.weaponIdx].description =
          action.payload.description;
        break;
      default:
    }
  },
);
