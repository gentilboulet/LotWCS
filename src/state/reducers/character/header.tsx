import { Draft, produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../../actions/character/header";
import { ICharacterState } from "../../models/character";
import { setRank } from "../../models/character/header";

export const headerReducer = produce(
  (draft: Draft<ICharacterState>, action: ActionType<typeof actions>) => {
    switch (action.type) {
      case getType(actions.setName):
        draft.name = action.payload.name;
        break;
      case getType(actions.setConcept):
        draft.concept = action.payload.concept;
        break;
      case getType(actions.setRank):
        setRank(draft, action.payload.rank);
        break;
      case getType(actions.setArchetype):
        draft.archetype = action.payload.archetype;
        break;
      default:
    }
  },
);
