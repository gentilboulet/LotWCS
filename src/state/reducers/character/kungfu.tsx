import { Draft, produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../../actions/character/kungfu";
import {
  addKungFuTechnique,
  IKungFuState,
  openStyle,
} from "../../models/character/kungfu";

export const kungfuReducer = produce(
  (draft: Draft<IKungFuState>, action: ActionType<typeof actions>) => {
    switch (action.type) {
      case getType(actions.openStyle):
        openStyle(draft, action.payload.kungfuType, action.payload.uid);
        break;
      case getType(actions.buyTechnique):
        addKungFuTechnique(
          draft,
          action.payload.kungfuType,
          action.payload.styleUid,
          action.payload.uid,
        );
        break;
      case getType(actions.customStyleName):
      case getType(actions.customTechniqueName):
      default:
    }
  },
);
