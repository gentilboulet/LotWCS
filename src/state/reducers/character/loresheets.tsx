import { Draft, produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../../actions/character/loresheets";
import {
  buyLoresheetOption,
  ILoresheetsState,
  openLoresheet,
} from "../../models/character/loresheets";

export const loresheetsReducer = produce(
  (draft: Draft<ILoresheetsState>, action: ActionType<typeof actions>) => {
    switch (action.type) {
      case getType(actions.open):
        openLoresheet(draft, action.payload.uid);
        break;
      case getType(actions.buyOption):
        buyLoresheetOption(
          draft,
          action.payload.lsUid,
          action.payload.uid,
          action.payload.payload,
        );
        break;
      default:
    }
  },
);
