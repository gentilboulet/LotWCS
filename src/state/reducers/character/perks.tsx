import { Draft, produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../../actions/character/perks";
import { ICharacterState } from "../../models/character";
import { applyBonuses } from "../../models/character/bonuses";
import { pushToDiscounts } from "../../models/character/discounts";

export const perksReducer = produce(
  (draft: Draft<ICharacterState>, action: ActionType<typeof actions>) => {
    switch (action.type) {
      case getType(actions.applyBonus):
        applyBonuses(draft, action.payload.bonuses);
        break;
      case getType(actions.pushDiscount):
        pushToDiscounts(draft, action.payload.discounts);
        break;
      default:
    }
  },
);
