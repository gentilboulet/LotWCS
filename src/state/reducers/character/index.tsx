import { produce, Draft } from "immer";
import { isActionOf } from "typesafe-actions";

import { ICharacterAction } from "../../actions/character";
import { ICharacterState } from "../../models/character";

// Sub Reducers
import { chiReducer } from "./chi";
import { costReducer } from "./costs";
import { headerReducer } from "./header";
import { kungfuReducer } from "./kungfu";
import { loresheetsReducer } from "./loresheets";
import { perksReducer } from "./perks";
import { skillsReducer } from "./skills";
import { virtuesReducer } from "./virtues";

import * as chi from "../../actions/character/chi";
import * as costs from "../../actions/character/costs";
import * as header from "../../actions/character/header";
import * as kungfu from "../../actions/character/kungfu";
import * as loresheets from "../../actions/character/loresheets";
import * as perks from "../../actions/character/perks";
import * as skills from "../../actions/character/skills";
import * as virtues from "../../actions/character/virtues";

export const globalReducer = produce(
  (draft: Draft<ICharacterState>, action: ICharacterAction) => {
    if (isActionOf(Object.values(chi), action)) {
      draft.chi = chiReducer(draft.chi, action);
    }
    if (isActionOf(Object.values(costs), action)) {
      costReducer(draft, action);
    }
    if (isActionOf(Object.values(header), action)) {
      headerReducer(draft, action);
    }
    if (isActionOf(Object.values(kungfu), action)) {
      draft.kungfu = kungfuReducer(draft.kungfu, action);
    }
    if (isActionOf(Object.values(loresheets), action)) {
      draft.loresheets = loresheetsReducer(draft.loresheets, action);
    }
    if (isActionOf(Object.values(perks), action)) {
      perksReducer(draft, action);
    }
    if (isActionOf(Object.values(skills), action)) {
      draft.skills = skillsReducer(draft.skills, action);
    }
    if (isActionOf(Object.values(virtues), action)) {
      draft.virtues = virtuesReducer(draft.virtues, action);
    }
  },
);
