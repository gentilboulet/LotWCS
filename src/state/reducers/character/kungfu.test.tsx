import { ActionType } from "typesafe-actions";

import * as data from "../../../data/kungfu";
import * as types from "../../../data/kungfu/types";
import { zeroCost } from "../../models/character/costs";

import * as actions from "../../actions/character/kungfu";
import { kungfuReducer } from "./kungfu";

import { ICharacterState } from "../../models/character";
import { testingStateFactory } from "../../models/character/testing";

const testingState: ICharacterState = testingStateFactory();

describe("Testing openStyle action", () => {
  it("should accept a new external style", () => {
    data.externalKungfu.forEach(style => {
      const newStyle = actions.openStyle(
        style.uid,
        types.KUNGFU_EXTERNAL,
        zeroCost,
      );
      expect(Object.keys(testingState.kungfu.KUNGFU_EXTERNAL).length).toBe(0);
      const result = kungfuReducer(testingState.kungfu, newStyle);
      expect(Object.keys(result.KUNGFU_EXTERNAL).length).toBe(1);
    });
  });

  it("should accept a new internal style", () => {
    data.internalKungfu.forEach(style => {
      const newStyle = actions.openStyle(
        style.uid,
        types.KUNGFU_INTERNAL,
        zeroCost,
      );
      expect(Object.keys(testingState.kungfu.KUNGFU_INTERNAL).length).toBe(0);
      const result = kungfuReducer(testingState.kungfu, newStyle);
      expect(Object.keys(result.KUNGFU_INTERNAL).length).toBe(1);
    });
  });
});

describe("Testing addStyleTechnique action", () => {
  it("should add a new external style technique", () => {
    data.externalKungfu.forEach(style => {
      const externalStyle = actions.openStyle(
        style.uid,
        types.KUNGFU_EXTERNAL,
        zeroCost,
      );
      const stateWithStyle = kungfuReducer(testingState.kungfu, externalStyle);
      style.techniques.forEach(tech => {
        const addTechnique = actions.buyTechnique(
          style.uid,
          tech.uid,
          types.KUNGFU_EXTERNAL,
          zeroCost,
        );

        expect(stateWithStyle.KUNGFU_EXTERNAL[style.uid].length).toBe(0);
        const stateWithTechnique = kungfuReducer(stateWithStyle, addTechnique);
        expect(stateWithTechnique.KUNGFU_EXTERNAL[style.uid].length).toBe(1);
      });
    });
  });

  it("should add a new internal style technique", () => {
    data.internalKungfu.forEach(style => {
      const internalKungfu = actions.openStyle(
        style.uid,
        types.KUNGFU_INTERNAL,
        zeroCost,
      );
      const stateWithStyle = kungfuReducer(testingState.kungfu, internalKungfu);
      style.techniques
        .filter(tech => tech.level > 1) // level 1 is free on style buy
        .forEach(tech => {
          const addTechnique = actions.buyTechnique(
            style.uid,
            tech.uid,
            types.KUNGFU_INTERNAL,
            zeroCost,
          );

          expect(stateWithStyle.KUNGFU_INTERNAL[style.uid].length).toBe(1);
          const stateWithTechnique = kungfuReducer(
            stateWithStyle,
            addTechnique,
          );
          expect(stateWithTechnique.KUNGFU_INTERNAL[style.uid].length).toBe(2);
        });
    });
  });
});

describe("Testing kungfuReducer", () => {
  it("should do nothing with a junk action", () => {
    const junk = { type: "JUNK_ACTION" };
    expect(
      kungfuReducer(testingState.kungfu, junk as ActionType<typeof actions>),
    ).toMatchSnapshot();
  });
});
