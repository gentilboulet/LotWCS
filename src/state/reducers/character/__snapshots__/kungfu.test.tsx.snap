// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing kungfuReducer should do nothing with a junk action 1`] = `
Object {
  "KUNGFU_EXTERNAL": Object {},
  "KUNGFU_INTERNAL": Object {},
  "noRestrictionInternal": Array [],
}
`;
